File: apps/api/src/analytics/overview.ts
import * as Clock from 'effect/Clock';
import * as Effect from 'effect/Effect';
import { Infrastructure } from '../platform/resources';
import { attempt, invalid, attemptSync } from '../shared/errors';
import { getEnvironment } from '../sites/service';
import { dateRange, report } from './reports';

export const overview = Effect.fn('overview')(function* (
  owner: string,
  site: string,
  key: string,
  query: Record<string, string>,
) {
  const r = yield* Infrastructure;
  const env = yield* getEnvironment(owner, site, key);

  const now = yield* Clock.currentTimeMillis,
    cutoff = new Date(now - 29 * 86400000).toISOString().slice(0, 10),
    values = ['path', 'referrer', 'country', 'device'].map((k) => query[k] ?? null);

  // eslint-disable-next-line no-control-regex -- Reject control characters at the privacy boundary.
  if (values.some((v) => v !== null && (v.length > 2048 || /[\x00-\x1f\x7f]/.test(v))))
    return yield* invalid('Invalid overview filter.');
  if (query.window && query.window !== '24h') return yield* invalid('Invalid overview window.');

  const hourly = query.window === '24h',
    filtered = values.some((v) => v !== null);

  const range = hourly
    ? yield* attemptSync(() =>
        dateRange({
          from: new Date(now - 86400000).toISOString().slice(0, 10),
          to: new Date(now).toISOString().slice(0, 10),
        }),
      )
    : yield* attemptSync(() => dateRange(query));

  const currentStart = hourly ? now - 86400000 : Date.parse(range.from),
    currentEnd = hourly ? now : Date.parse(range.to) + 86400000;

  const priorStart = currentStart - (currentEnd - currentStart),
    priorEnd = currentStart;

  const build = Effect.fn('build')(function* (start: number, end: number, full: boolean) {
    const r = yield* Infrastructure;

    const dates = {
      from: new Date(start).toISOString().slice(0, 10),
      to: new Date(end - 1).toISOString().slice(0, 10),
    };

    if (!hourly && !filtered) {
      const result: Record<string, unknown> = {
        overview: yield* report(owner, site, 'overview', { ...dates, environment: key }),
        timeseries: yield* report(owner, site, 'timeseries', { ...dates, environment: key }),
      };

      if (full)
        for (const dimension of ['path', 'referrer', 'country', 'device', 'event'])
          result[dimension] = yield* report(owner, site, 'breakdown', {
            ...dates,
            environment: key,
            dimension,
          });

      return result;
    }

    const boundsStart = hourly ? start : Math.max(start, Date.parse(cutoff));
    const cte = `WITH filtered AS MATERIALIZED (SELECT * FROM events WHERE site_id=$1 AND day BETWEEN $2 AND $3 AND received_at>=$8::timestamptz AND received_at<$9::timestamptz AND ($4::text IS NULL OR path=$4) AND ($5::text IS NULL OR referrer=$5) AND ($6::text IS NULL OR country=$6) AND ($7::text IS NULL OR device=$7))`;
    let extra = '';

    if (full)
      for (const dimension of ['path', 'referrer', 'country', 'device', 'event']) {
        const column = dimension === 'event' ? 'name' : dimension,
          type = dimension === 'event' ? 'event' : 'pageview';

        extra += `, '${dimension}',jsonb_build_object('data',coalesce((SELECT jsonb_agg(jsonb_build_object('value',value,'count',n) ORDER BY n DESC,value) FROM (SELECT ${column} AS value,count(*) AS n FROM filtered WHERE type='${type}' GROUP BY ${column} ORDER BY n DESC,value LIMIT 10) b),'[]'::jsonb))`;
      }

    const bucket = hourly
        ? "date_bin(interval '1 hour',received_at,$8::timestamptz)"
        : "day::timestamp AT TIME ZONE 'UTC'",
      series = hourly
        ? "generate_series($8::timestamptz,$9::timestamptz-interval '1 hour',interval '1 hour')"
        : "generate_series($2::date::timestamp AT TIME ZONE 'UTC',$3::date::timestamp AT TIME ZONE 'UTC',interval '1 day')";

    return yield* attempt(() =>
      r.analytics.begin(async (tx) => {
        await tx`SET TRANSACTION READ ONLY`;

        const [row] = await tx.unsafe(
          `${cte},buckets AS (SELECT ${bucket} AS at,count(*) FILTER(WHERE type='pageview') AS pageviews,count(*) FILTER(WHERE type='event') AS custom_events,count(DISTINCT(day,visitor)) AS visitors FROM filtered GROUP BY 1),series AS (SELECT d AS at,coalesce(pageviews,0) AS pageviews,coalesce(custom_events,0) AS custom_events,coalesce(visitors,0) AS visitors FROM ${series} d LEFT JOIN buckets ON buckets.at=d) SELECT jsonb_build_object('overview',jsonb_build_object('pageviews',coalesce(sum(pageviews),0),'customEvents',coalesce(sum(custom_events),0),'dailyUniqueVisitors',(SELECT count(DISTINCT(day,visitor)) FROM filtered)),'timeseries',jsonb_build_object('data',jsonb_agg(jsonb_build_object('day',(at AT TIME ZONE 'UTC')::date,${hourly ? "'at',at," : ''}'pageviews',pageviews,'customEvents',custom_events,'dailyUniqueVisitors',visitors) ORDER BY at))${extra})::text AS data FROM series`,
          [
            env.id,
            dates.from,
            dates.to,
            ...values,
            new Date(boundsStart).toISOString(),
            new Date(end).toISOString(),
          ],
        );

        return JSON.parse(row.data) as Record<string, unknown>;
      }),
    );
  });

  const current = yield* build(currentStart, currentEnd, true);

  const compare =
    hourly ||
    priorStart >=
      (filtered
        ? Date.parse(cutoff)
        : Date.parse(new Date(now).toISOString().slice(0, 10)) - 729 * 86400000);

  const previous = compare ? yield* build(priorStart, priorEnd, false) : null;

  const annotations = yield* attempt(
    () =>
      r.primary`SELECT id,day::text,label FROM overview_annotations WHERE environment_id=${env.id}::uuid AND day BETWEEN ${range.from} AND ${range.to} ORDER BY day,created_at,id`,
  );

  const definitions = yield* attempt(
    () =>
      r.primary`SELECT id,name,match_type AS "matchType",match_value AS "matchValue",created_at AS "createdAt" FROM conversion_goals WHERE environment_id=${env.id}::uuid ORDER BY created_at,id`,
  );

  const counts = yield* attempt(() =>
    r.analytics.begin(async (tx) => {
      await tx`SET TRANSACTION READ ONLY`;

      const params = [
        env.id,
        range.from,
        range.to,
        ...values,
        new Date(hourly ? currentStart : Math.max(currentStart, Date.parse(cutoff))).toISOString(),
        new Date(currentEnd).toISOString(),
      ];

      const cte = `WITH filtered AS MATERIALIZED (SELECT * FROM events WHERE site_id=$1 AND day BETWEEN $2 AND $3 AND received_at>=$8::timestamptz AND received_at<$9::timestamptz AND ($4::text IS NULL OR path=$4) AND ($5::text IS NULL OR referrer=$5) AND ($6::text IS NULL OR country=$6) AND ($7::text IS NULL OR device=$7))`;

      const rows = await tx.unsafe(
        `${cte} SELECT c.goal_id,count(*)::bigint AS conversions,count(DISTINCT(c.day,c.visitor))::bigint AS visitors FROM goal_conversions c JOIN filtered e ON e.id=c.event_id AND e.site_id=c.environment_id AND e.received_at=c.received_at GROUP BY c.goal_id`,
        params,
      );

      const [total] = await tx.unsafe(
        `${cte} SELECT count(DISTINCT(day,visitor))::bigint AS visitors FROM filtered`,
        params,
      );

      return { rows, visitors: Number(total.visitors) };
    }),
  );

  return {
    ...current,
    previous,
    previousRange: hourly
      ? { from: new Date(priorStart).toISOString(), to: new Date(priorEnd).toISOString() }
      : {
          from: new Date(priorStart).toISOString().slice(0, 10),
          to: new Date(priorEnd - 1).toISOString().slice(0, 10),
          days: range.days,
          timezone: 'UTC',
        },
    ...(hourly
      ? {
          window: {
            from: new Date(currentStart).toISOString(),
            to: new Date(currentEnd).toISOString(),
            interval: 'hour',
          },
        }
      : {}),
    filtered,
    retainedFrom: cutoff,
    partial: filtered && range.from < cutoff,
    annotations,
    goals: {
      visitors: counts.visitors,
      goals: definitions.map((goal: { id: string }) => {
        const row = counts.rows.find((row: { goal_id: string }) => row.goal_id === goal.id);

        return {
          ...goal,
          conversions: Number(row?.conversions ?? 0),
          visitors: Number(row?.visitors ?? 0),
        };
      }),
    },
  };
});

File: crates/api/src/analytics/overview.rs
use super::{
    Parameters, features, read_transaction,
    reports::{self, DateRange},
};
use crate::{AppState, error::ApiError};
use chrono::{DateTime, Duration, NaiveDate, SecondsFormat, Utc};
use serde_json::{Value, json};
use uuid::Uuid;

const FILTERED: &str = "WITH filtered AS MATERIALIZED (SELECT * FROM events WHERE site_id=$1 AND day BETWEEN $2 AND $3 AND received_at>=$8::timestamptz AND received_at<$9::timestamptz AND ($4::text IS NULL OR path=$4) AND ($5::text IS NULL OR referrer=$5) AND ($6::text IS NULL OR country=$6) AND ($7::text IS NULL OR device=$7))";

fn midnight(day: NaiveDate) -> Result<DateTime<Utc>, ApiError> {
    day.and_hms_opt(0, 0, 0)
        .map(|d| d.and_utc())
        .ok_or_else(ApiError::invalid)
}

fn timestamp(at: DateTime<Utc>) -> String {
    at.to_rfc3339_opts(SecondsFormat::Millis, true)
}

struct Window<'a> {
    state: &'a AppState,
    env: Uuid,
    hourly: bool,
    filtered: bool,
    values: [Option<&'a str>; 4],
    cutoff: NaiveDate,
}

impl Window<'_> {
    async fn build(
        &self,
        start: DateTime<Utc>,
        end: DateTime<Utc>,
        full: bool,
    ) -> Result<Value, ApiError> {
        let range = DateRange {
            from: start.date_naive(),
            to: (end - Duration::milliseconds(1)).date_naive(),
            days: 0,
        };
        if !self.hourly && !self.filtered {
            let query = range.parameters();
            let mut result = json!({
                "overview":reports::report(self.state,self.env,"overview",&query).await?,
                "timeseries":reports::report(self.state,self.env,"timeseries",&query).await?
            });
            if full {
                for dimension in ["path", "referrer", "country", "device", "event"] {
                    let mut query = query.clone();
                    query.insert("dimension".into(), dimension.into());
                    result[dimension] =
                        reports::report(self.state, self.env, "breakdown", &query).await?;
                }
            }
            return Ok(result);
        }
        let bounds_start = if self.hourly {
            start
        } else {
            start.max(midnight(self.cutoff)?)
        };
        let mut extra = String::new();
        if full {
            // Only fixed identifiers are interpolated; every request filter remains bound.
            for (dimension, column, kind) in [
                ("path", "path", "pageview"),
                ("referrer", "referrer", "pageview"),
                ("country", "country", "pageview"),
                ("device", "device", "pageview"),
                ("event", "name", "event"),
            ] {
                extra.push_str(&format!(", '{dimension}',jsonb_build_object('data',coalesce((SELECT jsonb_agg(jsonb_build_object('value',value,'count',n) ORDER BY n DESC,value) FROM (SELECT {column} AS value,count(*) AS n FROM filtered WHERE type='{kind}' GROUP BY {column} ORDER BY n DESC,value LIMIT 10) b),'[]'::jsonb))"));
            }
        }
        let (bucket, series, at) = if self.hourly {
            (
                "date_bin(interval '1 hour',received_at,$8::timestamptz)",
                "generate_series($8::timestamptz,$9::timestamptz-interval '1 hour',interval '1 hour')",
                "'at',at,",
            )
        } else {
            (
                "day::timestamp AT TIME ZONE 'UTC'",
                "generate_series($2::date::timestamp AT TIME ZONE 'UTC',$3::date::timestamp AT TIME ZONE 'UTC',interval '1 day')",
                "",
            )
        };
        let statement = format!(
            "{FILTERED},buckets AS (SELECT {bucket} AS at,count(*) FILTER(WHERE type='pageview') AS pageviews,count(*) FILTER(WHERE type='event') AS custom_events,count(DISTINCT(day,visitor)) AS visitors FROM filtered GROUP BY 1),series AS (SELECT d AS at,coalesce(pageviews,0) AS pageviews,coalesce(custom_events,0) AS custom_events,coalesce(visitors,0) AS visitors FROM {series} d LEFT JOIN buckets ON buckets.at=d) SELECT jsonb_build_object('overview',jsonb_build_object('pageviews',coalesce(sum(pageviews),0),'customEvents',coalesce(sum(custom_events),0),'dailyUniqueVisitors',(SELECT count(DISTINCT(day,visitor)) FROM filtered)),'timeseries',jsonb_build_object('data',jsonb_agg(jsonb_build_object('day',(at AT TIME ZONE 'UTC')::date,{at}'pageviews',pageviews,'customEvents',custom_events,'dailyUniqueVisitors',visitors) ORDER BY at)){extra}) FROM series"
        );
        let mut tx = read_transaction(self.state).await?;
        Ok(sqlx::query_scalar(&statement)
            .bind(self.env)
            .bind(range.from)
            .bind(range.to)
            .bind(self.values[0])
            .bind(self.values[1])
            .bind(self.values[2])
            .bind(self.values[3])
            .bind(bounds_start)
            .bind(end)
            .fetch_one(&mut *tx)
            .await?)
    }
}

pub(super) async fn read(
    state: &AppState,
    env: Uuid,
    query: &Parameters,
) -> Result<Value, ApiError> {
    let now = DateTime::from_timestamp_millis(Utc::now().timestamp_millis())
        .ok_or_else(ApiError::unavailable)?;
    let cutoff = (now - Duration::days(29)).date_naive();
    let values =
        ["path", "referrer", "country", "device"].map(|key| query.get(key).map(String::as_str));
    if values
        .iter()
        .flatten()
        .any(|v| v.encode_utf16().count() > 2048 || v.bytes().any(|b| b <= 31 || b == 127))
    {
        return Err(ApiError::invalid());
    }
    let window = query.get("window").filter(|v| !v.is_empty());
    if window.is_some_and(|v| v != "24h") {
        return Err(ApiError::invalid());
    }
    let hourly = window.is_some();
    let filtered = values.iter().any(Option::is_some);
    let range = if hourly {
        DateRange::parse(
            &[
                (
                    "from".into(),
                    (now - Duration::days(1)).date_naive().to_string(),
                ),
                ("to".into(), now.date_naive().to_string()),
            ]
            .into(),
        )?
    } else {
        DateRange::parse(query)?
    };
    let current_start = if hourly {
        now - Duration::days(1)
    } else {
        midnight(range.from)?
    };
    let current_end = if hourly {
        now
    } else {
        midnight(range.to)? + Duration::days(1)
    };
    let prior_start = current_start - (current_end - current_start);
    let prior_end = current_start;
    let builder = Window {
        state,
        env,
        hourly,
        filtered,
        values,
        cutoff,
    };
    let mut current = builder.build(current_start, current_end, true).await?;
    let compare = hourly
        || prior_start
            >= midnight(if filtered {
                cutoff
            } else {
                now.date_naive() - Duration::days(729)
            })?;
    current["previous"] = if compare {
        builder.build(prior_start, prior_end, false).await?
    } else {
        Value::Null
    };
    current["previousRange"] = if hourly {
        json!({"from":timestamp(prior_start),"to":timestamp(prior_end)})
    } else {
        json!({"from":prior_start.date_naive(),"to":(prior_end-Duration::days(1)).date_naive(),"days":range.days,"timezone":"UTC"})
    };
    if hourly {
        current["window"] =
            json!({"from":timestamp(current_start),"to":timestamp(current_end),"interval":"hour"});
    }
    current["filtered"] = json!(filtered);
    current["retainedFrom"] = json!(cutoff);
    current["partial"] = json!(filtered && range.from < cutoff);
    let mut tx = read_transaction(state).await?;
    let annotations:Vec<Value>=sqlx::query_scalar("SELECT jsonb_build_object('id',id,'day',day::text,'label',label) FROM overview_annotations WHERE environment_id=$1 AND day BETWEEN $2 AND $3 ORDER BY day,created_at,id")
        .bind(env).bind(range.from).bind(range.to).fetch_all(&mut *tx).await?;
    current["annotations"] = json!(annotations);
    let bounds_start = if hourly {
        current_start
    } else {
        current_start.max(midnight(cutoff)?)
    };
    let counts:Vec<Value>=sqlx::query_scalar(&format!("{FILTERED} SELECT jsonb_build_object('goal_id',c.goal_id,'conversions',count(*),'visitors',count(DISTINCT(c.day,c.visitor))) FROM goal_conversions c JOIN filtered e ON e.id=c.event_id AND e.site_id=c.environment_id AND e.received_at=c.received_at GROUP BY c.goal_id"))
        .bind(env).bind(range.from).bind(range.to).bind(values[0]).bind(values[1]).bind(values[2]).bind(values[3]).bind(bounds_start).bind(current_end).fetch_all(&mut *tx).await?;
    let visitors: i64 = sqlx::query_scalar(&format!(
        "{FILTERED} SELECT count(DISTINCT(day,visitor)) FROM filtered"
    ))
    .bind(env)
    .bind(range.from)
    .bind(range.to)
    .bind(values[0])
    .bind(values[1])
    .bind(values[2])
    .bind(values[3])
    .bind(bounds_start)
    .bind(current_end)
    .fetch_one(&mut *tx)
    .await?;
    current["goals"] = features::goals(&mut tx, env, counts, visitors).await?;
    Ok(current)
}
